contracts/RestrictedNORI.sol

```diff
@@ -0,0 +1,1089 @@
+      cumulativeQuantityToBurn += quantityToBurnForHolder;
+    }
+    quantitiesToBurnForHolders[tokenHolders.length() - 1] =
+      quantityToRevoke -
+      cumulativeQuantityToBurn;
+
+    for (uint256 i = 0; i < (tokenHolders.length()); i++) {
+      super._burn(
+        tokenHolders.at(i),
+        scheduleId,
+        quantitiesToBurnForHolders[i]
+      );
+      schedule.quantitiesRevokedByAddress[
+        tokenHolders.at(i)
+      ] += quantitiesToBurnForHolders[i];
+    }
+
+    schedule.totalQuantityRevoked += quantityToRevoke;
+    _bridgedPolygonNori.send(
+      // solhint-disable-previous-line check-send-result, because this isn't a solidity send
+      to,
+      quantityToRevoke,
+      ""
+    );
+    emit TokensRevoked(
+      block.timestamp, // solhint-disable-line not-rely-on-time, this is time-dependent
+      removalId,
+      scheduleId,
+      quantityToRevoke
+    );
+  }
+
+  /**
+   * Hook that is called before any token transfer. This includes minting and burning, as well as batched variants.
+   *
+   * @dev Follows the rules of hooks defined [here](
+   *  https://docs.openzeppelin.com/contracts/4.x/extending-contracts#rules_of_hooks)
+   * @dev See the ERC1155 specific version [here](
+   *  https://docs.openzeppelin.com/contracts/3.x/api/token/erc1155#ERC1155-_beforeTokenTransfer-address-address-address-uint256---uint256---bytes-)
+   *
+   * ##### Requirements:
+   *
+   * - the contract must not be paused
+   * - One of the following must be true:
+   *    - the operation is a mint (which should ONLY occur when BridgedPolygonNORI is being wrapped via `_depositFor`)
+   *    - the operation is a burn, which only happens during revocation and withdrawal:
+   *      - if the operation is a revocation, that permission is enforced by the TOKEN_REVOKER_ROLE
+   *      - if the operation is a withdrawal the burn amount must be <= the sender's claimable balance
+   *    - the operation is a transfer and _all_ of the following must be true:
+   *      - the operator is operating on their own balance (enforced in the inherited contract)
+   *      - the operator has sufficient balance to transfer (enforced in the inherited contract)
+   */
+  function _beforeTokenTransfer(
+    address operator,
+    address from,
+    address to,
+    uint256[] memory ids,
+    uint256[] memory amounts,
+    bytes memory data
+  )
+    internal
+    override(ERC1155PresetMinterPauserUpgradeable, ERC1155SupplyUpgradeable)
+  {
+    bool isMinting = from == address(0);
+    bool isBurning = to == address(0);
+
+    bool isWithdrawing = isBurning && from == operator;
+    bool isTransferring = !isMinting && !isBurning;
+
+    if (isBurning) {
+      for (uint256 i = 0; i < ids.length; i++) {
+        Schedule storage schedule = _scheduleIdToScheduleStruct[ids[i]];
+        schedule.releasedAmountFloor = _releasedBalanceOfSingleSchedule(ids[i]);
+      }
+    }
+    if (isWithdrawing) {
```

Comment:
can this be elseif?

Reply:
it can't actually because burning and withdrawing aren't mutually exclusive.  withdrawing involves burning, but not all burning is a withdrawal

Reply:
sorta makes me wonder if burning should just withdraw to the 0 address

Reply:
in the case where burning is a revocation, the account being withdrawn from (if we were to try to represent burning in this case as a withdrawal to the 0 address) would have to have enough claimable balance, but they might not necessarily have any claimable balance.

Reply:
why wouldn't we just leave that up to the evm to revert it? What's the point in us distinguishing here?

Reply:
oh, I think I understand now. Perhaps the better solution is to merge the loops though where we check if it's burning as part of the isWithdrawing  forloop? Not 100% sure if that results in a more efficient or inefficient version though, so double check my thinking if you think it doesnt make it better

Reply:
we can't put the check for `isBurning` inside the `isWithdrawing` for loop because there are burning cases that are not withdrawals.  I think you might mean to put the `isWithdrawing` check inside the `isBurning` block... because a withdrawal is a subset of a burn.  that gets us just one pass through the loop